charts/DataTable.tsx

```diff
@@ -1,61 +1,375 @@
 import * as React from "react"
-import { computed } from "mobx"
+import { computed, observable, action } from "mobx"
 import { observer } from "mobx-react"
+import classnames from "classnames"
+
+import { IconDefinition } from "@fortawesome/fontawesome-svg-core"
+import { FontAwesomeIcon } from "@fortawesome/react-fontawesome"
+import { faInfoCircle } from "@fortawesome/free-solid-svg-icons/faInfoCircle"
+import { faSortAmountDownAlt } from "@fortawesome/free-solid-svg-icons/faSortAmountDownAlt"
+import { faSortAmountUp } from "@fortawesome/free-solid-svg-icons/faSortAmountUp"
+import { faSortAlphaDown } from "@fortawesome/free-solid-svg-icons/faSortAlphaDown"
+import { faSortAlphaUpAlt } from "@fortawesome/free-solid-svg-icons/faSortAlphaUpAlt"
 
 import { ChartConfig } from "./ChartConfig"
-import { capitalize } from "./Util"
-import { DataTableTransform, DataTableRow } from "./DataTableTransform"
+import { capitalize, some, defaultTo, orderBy } from "./Util"
+import { Tippy } from "./Tippy"
+import {
+    DataTableTransform,
+    DataTableRow,
+    TargetYearMode,
+    SortOrder,
+    ColumnKey,
+    isSingleValue,
+    DataTableDimension,
+    DataTableColumn,
+    DimensionValue,
+    isRangeValue,
+    RangeValueKey,
+    SingleValueKey,
+    Value
+} from "./DataTableTransform"
 
-interface DataTableProps {
+export interface DataTableProps {
     chart: ChartConfig
 }
 
+export interface DataTableState {
+    sort: DataTableSortState
+}
+
+const ENTITY_DIM_INDEX = -1
+
+export type DimensionIndex = number
+
+export interface DataTableSortState {
+    dimIndex: DimensionIndex
+    columnKey: ColumnKey | undefined
+    order: SortOrder
+}
+
+const DEFAULT_SORT_STATE: DataTableSortState = {
+    dimIndex: ENTITY_DIM_INDEX,
+    columnKey: undefined,
+    order: SortOrder.asc
+}
+
+const columnNameByType: Record<ColumnKey, string> = {
+    single: "Value",
+    start: "Start",
+    end: "End",
+    delta: "Change",
+    deltaRatio: "Change (%)"
+}
+
+function inverseSortOrder(order: SortOrder): SortOrder {
+    return order === SortOrder.asc ? SortOrder.desc : SortOrder.asc
+}
+
 @observer
 export class DataTable extends React.Component<DataTableProps> {
+    @observable private storedState: DataTableState = {
+        sort: DEFAULT_SORT_STATE
+    }
+
+    @computed get tableState(): DataTableState {
+        return {
+            sort: this.sortState
+        }
+    }
+
+    @computed get sortState(): DataTableSortState {
+        let { dimIndex, columnKey, order } = {
+            ...DEFAULT_SORT_STATE,
+            ...this.storedState.sort
+        }
+
+        // If not sorted by entity, then make sure the index of the chosen column exists
+        if (dimIndex !== ENTITY_DIM_INDEX) {
```

Comment:
```suggestion
        if (dimIndex !== ENTITY_DIM_INDEX && this.transform.dimensions.length) {
```

This simple check seemingly fixes https://github.com/owid/owid-grapher/pull/319#issuecomment-592537632 (while still preserving the selected sort column on indicator change), but please have a look yourself since you know way better what this code is supposed to do.

Reply:
Thanks! It didn't occur to me that there can be 0 dimensions, but it happens while an indicator is loading.

Anyway, the goal of that logic is to never return invalid config, so I've done it a bit differently – just moved the `Math.min` line up. Not the cleanest, and the whole injection of the entity index at `-1` is ugly, but... I think it should be ok for a demo